datafusion/core/src/physical_plan/aggregates/priority_map.rs

```diff
@@ -0,0 +1,969 @@
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! A memory-conscious aggregation implementation that limits group buckets to a fixed number
+
+use crate::physical_plan::aggregates::group_values::primitive::HashValue;
+use crate::physical_plan::aggregates::{
+    aggregate_expressions, evaluate_group_by, evaluate_many, AggregateExec,
+    PhysicalGroupBy,
+};
+use crate::physical_plan::{RecordBatchStream, SendableRecordBatchStream};
+use ahash::RandomState;
+use arrow::datatypes::i256;
+use arrow::util::pretty::print_batches;
+use arrow_array::builder::PrimitiveBuilder;
+use arrow_array::cast::AsArray;
+use arrow_array::{
+    downcast_primitive, Array, ArrayRef, ArrowPrimitiveType, PrimitiveArray, RecordBatch,
+    StringArray,
+};
+use arrow_schema::{DataType, SchemaRef};
+use datafusion_common::DataFusionError;
+use datafusion_common::Result;
+use datafusion_execution::TaskContext;
+use datafusion_physical_expr::aggregate::utils::adjust_output_array;
+use datafusion_physical_expr::PhysicalExpr;
+use futures::stream::{Stream, StreamExt};
+use half::f16;
+use hashbrown::raw::RawTable;
+use itertools::Itertools;
+use log::{trace, Level};
+use std::cmp::Ordering;
+use std::fmt::{Debug, Formatter};
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+pub struct GroupedTopKAggregateStream {
+    partition: usize,
+    row_count: usize,
+    started: bool,
+    schema: SchemaRef,
+    input: SendableRecordBatchStream,
+    aggregate_arguments: Vec<Vec<Arc<dyn PhysicalExpr>>>,
+    group_by: PhysicalGroupBy,
+    map: Box<dyn LimitedHashTable>,
+    heap: Box<dyn LimitedHeap>,
+    limit: usize,
+}
+
+unsafe impl Send for GroupedTopKAggregateStream {}
+
+impl GroupedTopKAggregateStream {
+    pub fn new(
+        aggr: &AggregateExec,
+        context: Arc<TaskContext>,
+        partition: usize,
+        limit: usize,
+    ) -> Result<Self> {
+        let agg_schema = Arc::clone(&aggr.schema);
+        let group_by = aggr.group_by.clone();
+        let input = aggr.input.execute(partition, Arc::clone(&context))?;
+        let aggregate_arguments =
+            aggregate_expressions(&aggr.aggr_expr, &aggr.mode, group_by.expr.len())?;
+        let (val_field, desc) = aggr
+            .get_minmax_desc()
+            .ok_or_else(|| DataFusionError::Internal("Min/max required".to_string()))?;
+
+        let (expr, _) = &aggr.group_expr().expr()[0];
+        let kt = expr.data_type(&aggr.input().schema())?;
+        let map = new_map(limit, kt)?;
+
+        let vt = val_field.data_type().clone();
+        let heap = new_heap(limit, desc, vt)?;
```

Comment:
This factory and the one above it are now independent, making this 2*N instead of N^2 permutations.